doc/source/development/roadmap.rst

```diff
@@ -0,0 +1,153 @@
+as prototyped in `Fletcher <https://github.com/xhochy/fletcher>`__)
+or in the Apache Arrow C++ library.
+
+Apache Arrow Interoperability
+-----------------------------
+
+`Apache Arrow <https://arrow.apache.org>`__ is a cross-language development
+platform for in-memory data. The Arrow logical types are closely aligned with
+typical pandas use cases.
+
+We'd like to provide better-integrated support for Arrow memory and data types
+within pandas. This will let us take advantage of its I/O capabilities and
+provides for better interoperability with other languages and libraries
+using Arrow.
+
+Block Manager Rewrite
+---------------------
+
+We'd like to replace pandas current internal data structures (a collection of
+1 or 2-D arrays) with a simpler collection of 1-D arrays.
+
+Pandas internal data model is quite complex. A DataFrame is made up of
+one or more 2-dimension "blocks", with one or more blocks per dtype. This
+collection of 2-D arrays is managed by the BlockManager.
+
+The primary benefit of the BlockManager is improved performance on certain
+operations (construction from a 2D array, binary operations, reductions across the columns),
+especially for wide DataFrames. However, the BlockManager substantially increases the
+complexity and maintenance burden of pandas'.
+
+By replacing the BlockManager we hope to achieve
+
+* Substantially simpler code
+* Easier extensibility with new logical types
+* Better user control over memory use and layout
+* Improved microperformance
+* Option to provide a C / Cython API to pandas' internals
+
+See `these design documents <https://dev.pandas.io/pandas2/internal-architecture.html#removal-of-blockmanager-new-dataframe-internals>`__
+for more.
+
+Decoupling of Indexing and Internals
+------------------------------------
+
+The code for getting and setting values in pandas' data structures needs refactoring.
+In particular, a clear separation must be made between code that
+converts keys (e.g., the argument to ``DataFrame.loc``) to positions from code that uses
+uses these positions to get or set values. This is related to the proposed BlockManager rewrite.
+Currently, the BlockManager sometimes uses label-based, rather than position-based, indexing.
+We propose that it should only work with positional indexing, and the translation of keys
+to positions should be entirely done at a higher level.
+
+Indexing is a complicated API with many subtleties. This refactor will require care
+and attention. More details are discussed at
+https://github.com/pandas-dev/pandas/wiki/(Tentative)-rules-for-restructuring-indexing-code
+
+Numba-Accelerated Operations
+----------------------------
+
+[Numba](https://numba.pydata.org) is a JIT compiler for Python code. We'd like to provide
+ways for users to apply their own Numba-jitted functions within pandas' groupby and window
+contexts. This will improve the performance of user-defined-functions in these operations
+by staying within compiled code.
```

Comment:
Maybe stupid comment, but: is this not already possible?! If I pass a groupby a Numba-jitted function, won't it just work?

Reply:
Yes, but it won't take full advantage of it, as it will be seen a generic python function that is being applied. I think the idea is to have special handling of numba functions to reduce overhead (call it from compiled / jitted code instead of python for loop).

@TomAugspurger this could also mention `apply`/`map` (not in groupby context)?